### PR TITLE
Assignment-22 Commit to Staging

### PR DESCRIPTION
**Pull request by Sagar-Fexle-May2022**
State Trigger - Rollup
District Trigger - Rollup

### CODE REVIEW

File: src/classes/DistrictTriggerHandler.cls [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+            if(dist.State__c !=null){
+                mapOfStatusToUpdate.put(dist.State__c, new State__c(Id = dist.State__c,Total_District__c = 0));
+            }
+            if(oldDistrictMap !=null && dist.State__c!= oldDistrictMap.get(dist.Id).State__c){
+                mapOfStatusToUpdate.put(oldDistrictMap.get(dist.Id).State__c,
+                                        new State__c(Id = oldDistrictMap.get(dist.Id).State__c,Total_District__c = 0));
+            }
+              
+        }
+        if(!mapOfStatusToUpdate.isEmpty()){
```
Comment (sushilchoudhary7) [inline src/classes/DistrictTriggerHandler.cls:31]: Close this if condition at the end of this method.

Comment (Sagar-Fexle-May2022) [reply src/classes/DistrictTriggerHandler.cls:31]: Okay sir

File: src/classes/ContactTriggerHandler.cls [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+**/
+public with sharing class ContactTriggerHandler {
+    /**   Description :   Method synchronize Record
+    *     Created By  :   Sagar Nirwal
+    *     Arguments   :   List<State__c>,List<State__c>
+    *     Return      :   void
+    *     Revision log:   V_1.0 - Created  - Sagar Nirwal - 28/08/2022
+    *                     V_1.1 - Modified - Sagar Nirwal - 28/08/2022 - Details what you modified
+    **/
+    public static void synchronizeContactAccountPropertyRecord(List<Contact> newList) {
```
Comment (sushilchoudhary7) [inline src/classes/ContactTriggerHandler.cls:16]: Add proper null and size check in this method.

File: src/classes/CountryTriggerHandler.cls [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+                    oldCountryMap.get(cou.id).Total_District__c || cou.Total_States__c 
+                    != oldCountryMap.get(cou.id).Total_States__c )){   
+                    setOfCountryId.add(cou.Id);               
+                }
+            }
+            if(!setOfCountryId.isEmpty()){
+                List<State__c> listOfState = [SELECT Id, country__r.Id,country__r.Name,country__r.Total_States__c,
+                                             country__r.Total_District__c, Name
+                                             FROM State__c WHERE country__r.Name IN : setOfCountryId];
+                Map<Id, State__c> mapOfStatewithCountryId = new Map<Id, State__c>();
```
Comment (sushilchoudhary7) [inline src/classes/CountryTriggerHandler.cls:30]: Change map into Map<Id, List<State__c>> and use that map.

File: src/classes/CountryTriggerHandler.cls [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                        if(oldCountryMap != null && (cou.Total_District__c != oldCountryMap.get(cou.id).Total_District__c 
+                            || cou.Total_States__c != oldCountryMap.get(cou.id).Total_States__c )) {  
+                            Country wCountry = new Country();
+                            wCountry.countryName = mapOfStatewithCountryId.get(cou.Id).Country__r.Name;
+                            wCountry.countryCode = mapOfStatewithCountryId.get(cou.Id).Country__r.Country_Code__c;
+                            wCountry.states = new List<states>();
+                            wCountry.Total_Number_Of_Districts = (Integer)cou.Total_District__c;
+                            wCountry.Total_Number_Of_States = (Integer)cou.Total_States__c;
+
+                            for(State__c st : listOfState)
```
Comment (sushilchoudhary7) [inline src/classes/CountryTriggerHandler.cls:48]: Use mapOfStatewithCountryId this map after doing above mentioned changes.

File: src/classes/CountryTriggerHandler.cls [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                for(State__c st : listOfState){
+                    if(!mapOfStatewithCountryId.containsKey(st.Country__c)){
+                        mapOfStatewithCountryId.put(st.Country__c, st);
+                    }
+                }
+
+                if(!newCountryList.isEmpty()){
+                    for(Country__c cou : newCountryList){
+                        if(oldCountryMap != null && (cou.Total_District__c != oldCountryMap.get(cou.id).Total_District__c 
+                            || cou.Total_States__c != oldCountryMap.get(cou.id).Total_States__c )) {
```
Comment (sushilchoudhary7) [inline src/classes/CountryTriggerHandler.cls:40]: Just add setOfCountryId.contains(cou.Id) check.

File: src/classes/DistrictTriggerHandler.cls [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+            for(AggregateResult result :[SELECT Count(Id)numOfDis, State__c 
+                                        FROM District__c WHERE State__c in:mapOfStatusToUpdate.keySet() GROUP BY State__c]){
+                mapOfStatusToUpdate.get((Id)result.get('State__c')).Total_District__c = (Decimal)result.get('numOfDis');   
+            }
+        }
+        if(!mapOfStatusToUpdate.isEmpty()){
+            Update mapOfStatusToUpdate.values();
```
Comment (sushilchoudhary7) [inline src/classes/DistrictTriggerHandler.cls:38]: Add this line after 35 line no need to add same condition multiple time.

File: src/classes/StateTriggerHandler.cls [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+                    State wrapState = new State();
+                    wrapState.stateName = st.Name;
+                    wrapState.totalDistrict = st.Total_District__c;
+                    wrapState.districts = new List<Districts>();
+                    for(District__c dis: mapOfState.get(st.Id).Districts__r){
+                        Districts wrapDistrict = new Districts();
+                        wrapDistrict.districtName = dis.Name;
+                        wrapState.districts.add(wrapDistrict);
+                    }
+                String jString = JSON.serialize(wrapState);
```
Comment (sushilchoudhary7) [inline src/classes/StateTriggerHandler.cls:38]: Fix the indentation.

File: src/classes/StateTriggerHandler.cls [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+            for(AggregateResult agr:[SELECT Country__c , COUNT(Id)totalStates, SUM(Total_District__c)totalDistrict 
+                                    FROM State__c WHERE Id
+                                    IN:mapOfCountryUpdate.keySet() GROUP BY Country__c]){
+                mapOfCountryUpdate.put(String.valueOf(agr.get('Country__c')),new Country__c(Id = (Id)agr.get('Country__c'),
+                                        Total_States__c = (Decimal)agr.get('totalStates'),
+                                        Total_District__c = (Decimal)agr.get('totalDistrict')));
+            }
+        }
+        if(!mapOfCountryUpdate.isEmpty()){
+            update mapOfCountryUpdate.values();
```
Comment (sushilchoudhary7) [inline src/classes/StateTriggerHandler.cls:79]: add this line after 76 no need to add the same condition multiple times.

**Review by sushilchoudhary7 — CHANGES_REQUESTED** [review]
Work on changes as per discussion in the call.

File: src/classes/CountryTriggerHandler.cls [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+    *     Arguments   :   List,Map
+    *     Return      :   void
+    *     Revision log:   V_1.0 - Created  - Sagar Nirwal - 02/09/2022
+    *                     V_1.1 - Modified - Sagar Nirwal - 02/09/2022 - Details what you modified
+    **/
+    public static void countryJsonData(List<Country__c> newCountryList ,Map<Id,Country__c> oldCountryMap){
+
+        Set<Id> setOfCountryId = new Set<Id>();
+        for(Country__c cou : newCountryList){
+            if(oldCountryMap != null){
```
Comment (sushilchoudhary7) [inline src/classes/CountryTriggerHandler.cls:20]: Your code is running on the blank update as well. Fix this.

**Review by sushilchoudhary7 — CHANGES_REQUESTED** [review]
Do changes as per discussion in the call.

File: src/classes/ContactTriggerHandler.cls [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    *     Created By  :   Sagar Nirwal
+    *     Arguments   :   List<State__c>,List<State__c>
+    *     Return      :   void
+    *     Revision log:   V_1.0 - Created  - Sagar Nirwal - 28/08/2022
+    *                     V_1.1 - Modified - Sagar Nirwal - 28/08/2022 - Details what you modified
+    **/
+    public static void synchronizeContactAccountPropertyRecord(List<Contact> newContactList) {
+        Set<Id> setOfAccountIds = new Set<Id>();
+        for(Contact con: newContactList){
+            setOfAccountIds.add(con.AccountId);
```
Comment (sushilchoudhary7) [inline src/classes/ContactTriggerHandler.cls:19]: This code is working in the blank update. Fix it.

File: src/classes/DistrictTriggerHandler.cls [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    /**   Description :   Method finalizing Count Of District On States
+    *     Created By  :   Sagar Nirwal
+    *     Arguments   :   List<State__c>,Map<Id, District__c>
+    *     Return      :   void
+    *     Revision log:   V_1.0 - Created  - Sagar Nirwal - 28/08/2022
+    *                     V_1.1 - Modified - Sagar Nirwal - 28/08/2022 - Details what you modified
+    **/
+    public static void finalizingCountOfDistrictOnStates(List<District__c> newDistrictList, Map<Id, District__c> oldDistrictMap){
+        
+        Map<Id, State__c> mapOfStatusToUpdate = new Map<Id, State__c>();
```
Comment (sushilchoudhary7) [inline src/classes/DistrictTriggerHandler.cls:18]: Change variable name to mapOfStatesToUpdate

File: src/classes/ContactTriggerHandler.cls [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    *     Created By  :   Sagar Nirwal
+    *     Arguments   :   List<Contact>
+    *     Return      :   void
+    *     Revision log:   V_1.0 - Created  - Sagar Nirwal - 28/08/2022
+    *                     V_1.1 - Modified - Sagar Nirwal - 28/08/2022 - Details what you modified
+    **/
+    public static void synchronizeContactAccountPropertyRecord(List<Contact> newContactList) {
+        Set<Id> setOfAccountIds = new Set<Id>();
+        for(Contact con: newContactList){
+            if(con.Status__c == null){
```
Comment (sushilchoudhary7) [inline src/classes/ContactTriggerHandler.cls:19]: And blank check here.

File: src/classes/ContactTriggerHandler.cls [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+                                                                FROM Account 
+                                                                WHERE Id IN: setOfAccountIds]);
+            System.debug(mapOfAccount);                                                            
+            for(Contact con: newContactList){
+                if(mapOfAccount.containskey(con.AccountId)) {
+                    Account acc = mapOfAccount.get(con.AccountId);
+                    if(acc.Status__c != null){
+                        con.Status__c = acc.Status__c;
+                        System.debug('Contact Status:======== ' + con.Status__c); 
+                    }else if(acc.Status__c == null){
```
Comment (sushilchoudhary7) [inline src/classes/ContactTriggerHandler.cls:37]: acc.Property__c != null add this condition remove existing condition from else if

File: src/classes/ContactTriggerHandler.cls [hunk 14]
```diff
@@ -0,0 +1,47 @@
...
+    *     Created By  :   Sagar Nirwal
+    *     Arguments   :   List<Contact>
+    *     Return      :   void
+    *     Revision log:   V_1.0 - Created  - Sagar Nirwal - 28/08/2022
+    *                     V_1.1 - Modified - Sagar Nirwal - 28/08/2022 - Details what you modified
+    **/
+    public static void synchronizeContactAccountPropertyRecord(List<Contact> newContactList) {
+        Set<Id> setOfAccountIds = new Set<Id>();
+        for(Contact con: newContactList){
+            if(con.Status__c != null || String.isBlank(con.Status__c)){
```
Comment (sushilchoudhary7) [inline src/classes/ContactTriggerHandler.cls:19]: If should be like if(con.Status__c == null && (oldMap == null || oldMap.get(con.Id).Status__c != con.Statuc__c))

File: src/classes/ContactTriggerHandler.cls [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+**/
+public with sharing class ContactTriggerHandler {
+    /**   Description :   Method synchronize Record
+    *     Created By  :   Sagar Nirwal
+    *     Arguments   :   List<Contact>
+    *     Return      :   void
+    *     Revision log:   V_1.0 - Created  - Sagar Nirwal - 28/08/2022
+    *                     V_1.1 - Modified - Sagar Nirwal - 28/08/2022 - Details what you modified
+    **/
+    public static void synchronizeContactAccountPropertyRecord(List<Contact> newContactList) {
```
Comment (sushilchoudhary7) [inline src/classes/ContactTriggerHandler.cls:16]: This method will run on insert and update both.

**Review by sushilchoudhary7 — APPROVED** [review]

**Pull request merged**